File: Behavioral/Observer/DIY.java
package Observer;

import java.util.ArrayList;
import java.util.List;

public class DIY {
    public static void main(String[] args) {
        var data = new DataSource();
        var canvas1 = new Canvas(data);
        var canvas2= new Canvas(data);
        var canvas3 = new Canvas(data);

        data.addObserver(canvas1);
        data.addObserver(canvas2);
        data.removeObserver(canvas3);

        data.setValue(
            "Drawing"
        );;
    }
    
}

class DataSource extends Subject{
    private String value;

    public String getValue(){
        return value;
    }

    public void setValue(String value){
        this.value=value;
        notifyObservers();
    }
}

class Subject{
    private List<Observer> observers = new ArrayList<>();
    
    public void addObserver(Observer observer){
        observers.add(observer);
    }

    public void removeObserver(Observer observer){
        observers.remove(observer);
    }

    public void notifyObservers(){
        for(var observer: observers){
           observer.update(); 
        }
    } 
}

class Canvas implements Observer{
    private DataSource data;
    public Canvas(DataSource data){
        this.data=data;
    }
    @Override
    public void update() {
      System.out.println("Canvas got updated:" + data.getValue() );
    }
    
}
interface Observer{
    void update();
}

File: TS/Behavioral/Observer/DIY.ts
interface IObservableDIY {
  addObserver(observerDIY: ObserverDIY): void;
  removeObserver(observerDIY: ObserverDIY): void;
  notify(...args: unknown[]): void;
}

interface ObserverDIY {
  update(): void;
  notify(...args: unknown[]): void;
}

class ObserverSubject implements IObservableDIY {
  private observers: ObserverDIY[] = [];

  addObserver(observerDIY: ObserverDIY): void {
    this.observers.push(observerDIY);
  }

  removeObserver(observerDIY: ObserverDIY): void {
    const index = this.observers.indexOf(observerDIY);
    if (index > -1) {
      this.observers.splice(index, 1);
    }
  }

  notify(...args: unknown[]): void {
    this.observers.forEach((observerDIY) => {
      observerDIY.notify(...args);
      observerDIY.update();
    });
  }
}

class DataSource extends ObserverSubject {
  private value: string;

  getValue(): string {
    return this.value;
  }

  setValue(value: string): void {
    this.value = value;
    this.notify(value);
  }
}
class Canvas implements ObserverDIY {
  private data: DataSource;

  constructor(data: DataSource) {
    this.data = data;
  }
  update(): void {
    console.log("Canvas got updated: " + this.data.getValue());
  }
  notify(...args: unknown[]): void {
    console.log(`Canvas set: ${JSON.stringify(args)}`);
  }
}

// The Client
const data = new DataSource();
const canvas1 = new Canvas(data);
const canvas2 = new Canvas(data);
const canvas3 = new Canvas(data);

data.addObserver(canvas1);
data.addObserver(canvas2);
data.removeObserver(canvas3);

data.setValue("Drawing");
